ZenPacks/zenoss/OpenStackInfrastructure/datamaps.py

```diff
@@ -0,0 +1,201 @@
+        # for convenience, ignore None objmaps if passed in.  (mapping functions
+        # can return None if there's no data to apply, and it's easier to just
+        # check for that here)
+        if objmap is None:
+            return
+
+        component_id = objmap.id
+
+        if component_id not in self.held_objmaps:
+            self.held_objmaps[component_id] = (self.now(), objmap)
+        else:
+            _, stored_objmap = self.held_objmaps[component_id]
+
+            # Only overwrite _add and remove if the value is True.
+            if getattr(objmap, '_add', False):
+                stored_objmap._add = objmap._add
+            if getattr(objmap, 'remove', False):
+                stored_objmap.remove = objmap.remove
+
+            # update the existing objmap's other values
+            for p in objmap.__dict__:
+                if p not in ('_attrs', 'id', 'modname', 'relname', '_add', 'remove'):
+                    setattr(stored_objmap, p, getattr(objmap, p))
+
+    def _expire_blacklisted_components(self):
+        now = self.now()
+        expired = set()
+        for component_id, added_time in self.blacklisted_components.iteritems():
+            age = now - added_time
+            if age > self.delete_blacklist_seconds:
+                expired.add(component_id)
+        for component_id in expired:
+            del self.blacklisted_components[component_id]
+
+    def _new_objmap(self, objmap):
+        # Create a new empty objmap based on the suppied one, with all
+        # structural properties carried over, but no data, no _add, and no
+        # remove.
+        new_objmap = ObjectMap()
+
+        if objmap.modname:
+            new_objmap.modname = objmap.modname
+        if objmap.compname:
+            new_objmap.compname = objmap.compname
+        if objmap.classname:
+            new_objmap.classname = objmap.classname
+        if hasattr(objmap, 'relname'):
+            new_objmap.relname = objmap.relname
+        if hasattr(objmap, 'id'):
+            new_objmap.id = objmap.id
+
+        return new_objmap
+
+    def drain(self):
+        # remove entries from self.blacklisted_components that are too old.
+        self._expire_blacklisted_components()
+
+        # return a list of objectmaps that are ready to be applied.
+        released_objmaps = {}
+        now = self.now()
+
+        # Blacklist any components that have been both added and deleted
+        # while in this queue- no need to send such updates to zodb, after all.
+        for component_id in self.held_objmaps:
+            first_update, objmap = self.held_objmaps[component_id]
+            if getattr(objmap, 'remove', False) and getattr(objmap, '_add', False):
+                self.blacklisted_components[component_id] = now
+
+        # Remove any updates pertaining to components which are
+        # blacklisted.
+        for component_id in self.blacklisted_components:
+            del self.held_objmaps[component_id]
```

Comment:
This is probably fine, but it looks like a KeyError wating to happen.

Reply:
in fact it was, and i fixed locally it just before your comment came in!